src/bigtable/mod.rs

```diff
@@ -0,0 +1,537 @@
+        // to reduce re-querying. See the go implementation at
+        // https://github.com/googleapis/google-cloud-go/blob/21fc86d799bcdffc3db6f697b30ca48e072e7044/bigtable/bigtable.go#L171
+        let row_stream = response
+            .scan::<RowInProgress, Result<Vec<Result<Row, tonic::Status>>, tonic::Status>, _, _>(
+                RowInProgress::default(),
+                move |state, message| {
+                    let message = match message {
+                        Ok(m) => m,
+                        Err(e) => return futures::future::ready(Some(Err(e))),
+                    };
+
+                    for chunk in message.chunks {
+                        if !chunk.row_key.is_empty() {
+                            // We don't need to check if there's an existing row to store, because
+                            // RowStatus tells us when to do that.
+                            state.row.key = chunk.row_key;
+                        }
+
+                        if let Some(family_name) = chunk.family_name {
+                            state.new_family(family_name);
+                        }
+
+                        if let Some(qualifier) = chunk.qualifier {
+                            state.new_column(qualifier.into());
+                        }
+
+                        if Some(chunk.timestamp_micros)
+                            != state.last_cell_mut().map(|cell| cell.timestamp_micros)
+                        {
+                            state.new_cell(chunk.timestamp_micros, chunk.value_size);
+                        }
+
+                        // TODO: in the (presumably reasonably common case that the buffer comes
+                        // in a single chunk, we should avoid copying the data.
+                        state.value.extend_from_slice(&chunk.value);
+
+                        if let Some(row_status) = chunk.row_status {
+                            let row = state.finish_row();
+                            if matches!(
+                                row_status,
+                                v2::read_rows_response::cell_chunk::RowStatus::CommitRow(_),
+                            ) {
+                                state.completed.push(Ok(row));
+                            }
+                        }
+                    }
+
+                    let ret = std::mem::take(&mut state.completed);
+                    futures::future::ready(Some(Ok(ret)))
+                },
+            )
+            .flat_map(|rows| match rows {
+                Ok(rows) => futures::stream::iter(rows),
+                Err(e) => futures::stream::iter(vec![Err(e)]),
+            });
+
+        Ok(RowStream {
+            inner: Box::pin(row_stream),
+        })
+    }
+
+    /// Queries a range of rows from a table, returning just the keys.
+    pub async fn read_row_range_keys(
+        &mut self,
+        table_name: &str,
+        range: impl RangeBounds<Bytes>,
+        rows_limit: i64,
```

Comment:
How do you feel about using `Option<u64>` for arguments like `rows_limit` and then converting to i64 (returning an error if the value exceeds the range)? Then `None` is more clear to say there's no limit than `0`